lib/server/services/qualityControlFlag/QcFlagService.js

```diff
@@ -53,6 +56,192 @@ class QcFlagService {
+     * @param {Partial<QcFlag>} parameters flag instance parameters
+     * @param {object} [relations] QC Flag Type entity relations
+     * @param {Partial<UserIdentifier>} [relations.user] user identifiers
+     * @param {number} [parameters.flagTypeId] flag type id
+     * @param {number} [parameters.runNumber] associated run's number
+     * @param {number} [parameters.dataPassId] associated dataPass' id
+     * @param {number} [parameters.dplDetectorId] associated dplDetector's id
+     * @return {Promise<QcFlag>} promise
+     * @throws {BadParameterError, NotFoundError}
+     */
+    async createForDataPass(parameters, relations = {}) {
+        const {
+            from: fromTime = null,
+            to: toTime = null,
+            comment,
+        } = parameters;
+        const {
+            user: { userId, externalUserId } = {},
+            flagTypeId,
+            runNumber,
+            dataPassId,
+            dplDetectorId,
+        } = relations;
+
+        if (fromTime && toTime && !(fromTime < toTime)) {
+            throw new BadParameterError('Parameter "to" timestamp must be greater than "from" timestamp');
+        }
+        return dataSource.transaction(async () => {
+            // Check user
+            const user = await getUserOrFail({ userId, externalUserId });
+
+            // Check associations
+            const dplDetector = await DplDetectorRepository.findOne({ where: { id: dplDetectorId } });
+            if (!dplDetector) {
+                throw new BadParameterError(`No DPL Detector with given id (${dplDetectorId})`);
+            }
+            if (!dplDetector.name || NON_QC_DETECTORS.has(dplDetector.name)) {
+                throw new BadParameterError(`Bad DPL Detector selection, QC flag creation for "${dplDetector.name}" is impossible`);
+            }
+
+            const targetRun = await RunRepository.findOne({
+                subQuery: false,
+                attributes: ['timeTrgStart', 'timeTrgEnd'],
+                where: { runNumber },
+                include: [
+                    {
+                        association: 'dataPass',
+                        where: { id: dataPassId },
+                        through: { attributes: [] },
+                        attributes: ['id'],
+                        required: true,
+                    },
+                    {
+                        association: 'detectors',
+                        where: { name: dplDetector.name },
+                        through: { attributes: [] },
+                        attributes: [],
+                        required: true,
+                    },
+                ],
+            });
+            if (!targetRun) {
+                // eslint-disable-next-line max-len
+                throw new BadParameterError(`You cannot insert flag for data pass (id:${dataPassId}), run (runNumber:${runNumber}), detector (name:${dplDetector.name}) as there is no association between them`);
```

Comment:
Try to be uniform, for example before you put values in "", sometimes we put <>, here you use ()...

Reply:
refactored